### PR TITLE
Wallet and Transaction Component 

### PR DESCRIPTION
**Pull request by codyenokida**
Hi all!

PR to view recent transactions and view wallet status;

Features:
1. Expand transaction list from wallet widget
2. persist 3 most recent transactions
3. updated storybook (although it looks wonky on storybook)
4. added timestamp to all transactions (when loading, may need to remove from old transaction struct from localStorage)
5. new logic to metadata subscriber to fix bug that persisted transactions across wallet changes
6. updated styling to match figma

Some TODOs that may change in future PRs:
1. background color of widget (don't know the color for light mode.. probably will be the same as the side content bar, but that's tbd at the moment)
2. Fixed length of wallet widget?

Comments/feedback appreciated!

### CODE REVIEW

File: src/components/WalletButton/TransactionRow.styles.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import styled from "styled-components";
+
+export const Container = styled.div`
+  display: grid;
+  grid-auto-flow: column;
+  grid-template-columns: 0.5rem auto auto 0.5rem;
```
Comment (piersss) [inline src/components/WalletButton/TransactionRow.styles.tsx:6]: `0.5rem auto 0.5rem` and remove that span

Comment (piersss) [inline src/components/WalletButton/TransactionRow.styles.tsx:6]: `0.5rem auto 0.5rem` and remove that span

File: src/components/WalletButton/TransactionRow.tsx [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  if (type === "Order") {
+    const tx: SubmittedOrder = transaction as SubmittedOrder;
+    return (
+      <Container>
+        {tx.status === "succeeded" ? (
+          <HiOutlineCheck />
+        ) : tx.status === "processing" ? (
+          <RiLoader2Fill />
+        ) : (
+          <HiX />
```
Comment (piersss) [inline src/components/WalletButton/TransactionRow.tsx:71]: Add these icons to Icon component

File: src/components/WalletButton/TransactionRow.tsx [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  } else if (transaction.type === "Approval") {
+    const tx: SubmittedApproval = transaction as SubmittedApproval;
+    return (
+      <Container>
+        {tx.status === "succeeded" ? (
+          <HiOutlineCheck />
+        ) : tx.status === "processing" ? (
+          <RiLoader2Fill />
+        ) : (
+          <HiX />
```
Comment (piersss) [inline src/components/WalletButton/TransactionRow.tsx:125]: Duplication of code. Make a WalletTransactionIcon component or something.

File: src/components/WalletButton/TransactionRow.tsx [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            </>
+          )}
+        </TextContainer>
+        <SpanTitle></SpanTitle>
+        <Link
+          target="_blank"
+          rel="noreferrer"
+          href={`${getEtherscanURL(`${chainId}`, tx.hash)}`}
+        >
+          <MdOpenInNew />
```
Comment (piersss) [inline src/components/WalletButton/TransactionRow.tsx:112]: Add this icon to the Icon component as well.

File: src/components/WalletButton/TransactionRow.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            </SpanSubtitle>
+          </>
+        )}
+        <Link
+          target="_blank"
+          rel="noreferrer"
+          href={`${getEtherscanURL(`${chainId}`, tx.hash)}`}
+        >
+          <MdOpenInNew />
+        </Link>
```
Comment (piersss) [inline src/components/WalletButton/TransactionRow.tsx:149]: Duplication of code. Make a WalletTransactionLink component.

File: src/components/WalletButton/TransactionRow.tsx [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+      </Container>
+    );
+  }
+  return <div></div>;
```
Comment (piersss) [inline src/components/WalletButton/TransactionRow.tsx:153]: This is unnecessary.

```
if (type === "Order") {
   return ...
}

return ... (this is "Approval" then)
```

File: src/components/WalletButton/WalletButton.styles.tsx [hunk 7]
```diff
@@ -1,6 +1,90 @@
...
+  width: 100%;
+  z-index: 100;
+  padding: 0 1rem 1rem 1rem;
+  background-color: ${(props) =>
+    props.theme.name === "dark" ? props.theme.colors.darkGrey : "#F4F4F4"};
   display: flex;
+  flex-direction: row;
+  flex-wrap: wrap;
   align-items: center;
+  justify-content: center;
```
Comment (piersss) [inline src/components/WalletButton/WalletButton.styles.tsx:51]:
```
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
  position: absolute;
  width: 100%;
  padding: 0 1rem 1rem 1rem;
  z-index: 100;
  background-color: ${(props) =>
    props.theme.name === "dark" ? props.theme.colors.darkGrey : "#F4F4F4"};
```

File: src/components/WalletButton/WalletButton.tsx [hunk 8]
```diff
@@ -29,37 +47,111 @@ export type WalletButtonProps = {
...
+  tokens: TokenInfo[];
 };
 
 export const WalletButton = ({
   address,
-  className,
   isConnecting,
   onConnectWalletClicked,
   onDisconnectWalletClicked,
+  transactions,
```
Comment (piersss) [inline src/components/WalletButton/WalletButton.tsx:69]: `transactions = []`

File: src/components/WalletButton/WalletButton.tsx [hunk 9]
```diff
@@ -29,37 +47,111 @@ export type WalletButtonProps = {
...
+              <ExitButton onClick={() => setWalletOpen(!walletOpen)}>
+                <RiCloseLine />
+              </ExitButton>
+            )}
+          </StyledWalletButton>
+          {walletOpen && (
+            <WalletExtension>
+              <Line />
+              <TransactionContainer flex={transactions.length === 0}>
+                {transactions.length > 0 ? (
```
Comment (piersss) [inline src/components/WalletButton/WalletButton.tsx:103]: Remove this line, if you slice an empty array it's still empty. `[].slice(0, 3);`

File: src/components/WalletButton/WalletButton.tsx [hunk 10]
```diff
@@ -29,37 +47,111 @@ export type WalletButtonProps = {
...
+                seed={address}
+                bgColor="black"
+                color="#2b72ff"
+              />
+              <GreenCircle />
+            </BlockiesContainer>
+            <InfoHeading>{truncateEthAddress(address)}</InfoHeading>
+            {walletOpen && (
+              <ExitButton onClick={() => setWalletOpen(!walletOpen)}>
+                <RiCloseLine />
```
Comment (piersss) [inline src/components/WalletButton/WalletButton.tsx:95]: Add to Icon component.

File: src/components/WalletButton/WalletButton.tsx [hunk 11]
```diff
@@ -29,37 +47,111 @@ export type WalletButtonProps = {
...
+                        <TransactionRow
+                          transaction={transaction}
+                          approvalToken={token}
+                          type={transaction.type}
+                          chainId={chainId!}
+                          key={transaction.hash}
+                        />
+                      );
+                    }
+                    return null;
```
Comment (piersss) [inline src/components/WalletButton/WalletButton.tsx:139]: There's too much logic in your render. I would make a useEffect with transactions as dependency that generates an array of transactions based of TransactionRowProps. This way your render is way cleaner.

Also what's the return null at the end for? :p

File: src/helpers/getTimeBetweenTwoDates.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
+const MS_PER_MINUTE = 60000;
+const MS_PER_HOUR = 3600000;
+const MS_PER_DAY = 86400000;
+const MS_PER_MONTH = 2592000000;
+const MS_PER_YEAR = 31536000000;
+
+export default function getTimeBetweenTwoDates(date: Date) {
```
Comment (piersss) [inline src/helpers/getTimeBetweenTwoDates.ts:7]: `export default function getTimeBetweenTwoDates(date: Date): string {`

**Review by piersss — CHANGES_REQUESTED** [review]
Hey Cody. Nice progress far. Here's some feedback.

- Double check design. Font-sizes, font-weights, spacings, and the width of the wallet should be 344px, yours is 294px.

Screenshot 2021-08-13 at 20 56 18

- Timestamps show up as undefined for me.
- I think the Wallet should be next to the Sidebar.
- Component should be named Wallet, not WalletButton.
- TransactionRow should be named WalletTransaction.
- You can fix your component in storybook by wrapping it in a container with a max-with. See TokenSelection story for example.

File: src/components/WalletButton/WalletButton.tsx (hunk 9 above)
Comment (codyenokida) [reply src/components/WalletButton/WalletButton.tsx:103]: This line is left so that we can render "no transaction" if transactions is empty.

Comment (piersss) [reply src/components/WalletButton/WalletButton.tsx:103]: Ah yes I didn't notice.

**Review by piersss — APPROVED** [review]
Looks good! Just the styling is not close enough to the design. I think it's ok to merge for now and then I can do some tweaks to the styling.

**Pull request merged**